### DISCUSSION TITLE
SPI Flash Bootloader update

### DISCUSSION THREAD

**Discussion by AdityaICVS** [rauc/rauc#1496] [General]
Hello,

We have integrated RAUC in our project and its working fine so far. Now a new requirement has come-up which requires the bootloader to be placed in SPI flash instead of EMMC or SD-card.

I wanted to know whether that would be possible to do it through RAUC.

**Comment by jluebbe** [comment]
Take a look at the PR https://github.com/rauc/rauc/pull/1244 and the discussion there: atomic bootloader updates in SPI flash are not as simple as they may seem.

If (and how) atomic bootloader updates can be done depends on the SoC and the details of it's boot process. Which SoC are you using?

**Comment by ejoerns** [comment]
@AdityaICVS Where you place the bootloader is not really relevant for RAUC if you just use it to boot the system.

If you want to be able to update the bootloader with RAUC, then you must have a way to do it atomically.
For NOR, the options you have depend on how the SoC's ROM loader reads the NOR and if it supports falling back to other regions.

In RAUC there is no standard support for atomic bootloader updates in NOR yet, but there is https://github.com/rauc/rauc/pull/1244.

**Comment by AdityaICVS** [comment]
Hello,

Thanks for the response. 
I meant to ask about **updating** the bootloader in SPI flash through the RAUC. I don't see any provision to do it in the system.conf. So thought of checking whether it is possible at all.
Any upcoming plans for this? The SOC which I am using is imx8mp. It supports boot from sdcard/emmc/spi flash. Going ahead we want to keep the bootloader in SPI flash and looking for provision to update the same if required.

**Reply by ejoerns** [reply]
> Thanks for the response. I meant to ask about **updating** the bootloader in SPI flash through the RAUC.

Thanks for the clarification. Properly updating the bootloader with RAUC would mean to have a way to do it *atomically*, i.e. not only overwriting the existing bootloader but writing it to a different locations and switching to this location afterwards, like described in https://rauc.readthedocs.io/en/latest/advanced.html#updating-the-bootloader.

> I don't see any provision to do it in the system.conf. So thought of checking whether it is possible at all.

Non-atomic updates of NOR MTD partitions are possible and would work in RAUC already, but I would strongly discourage doing so.

> Any upcoming plans for this? The SOC which I am using is imx8mp. It supports boot from sdcard/emmc/spi flash. Going ahead we want to keep the bootloader in SPI flash and looking for provision to update the same if required.

For the upcoming plans, we pointed to #1244 which started implementing a mechanism for atomic updates in NOR.

If I remember correctly, we also have a project here with i.mx8mp where we would like to support atomic updates of a bootloader in NOR somewhere in the future, but there is no current work on it from our side.

Does that answer your questions?
